core/market/decentralized/src/lib.rs

```diff
@@ -22,3 +23,8 @@ pub mod migrations {
     #[derive(diesel_migrations::EmbedMigrations)]
     struct _Dummy;
 }
+
+/// These exports are expected to be used only in tests.
+pub mod testing {
+    pub use super::db::dao::{DemandDao, OfferDao};
```

Comment:
Rust tests are devided into unit tests that have access to everything in module and must be written in the same file, and integration tests (that have access to public interface only).

We want to test integration, but we need to dive into internals as well, so it's impossible to use only public interface. As example I can give testing incorrect behavior in discovery api - it is impossible to test incorrect behavior using market interface, if we implement market to work correctly.

Exposing stuff for tests worries me as well :(
I asked internet and it was the cleanest way to expose things. I considered rust config macros to make these things visible only in tests, but rust doesn't recompiles modules for tests, so it doesn't work.
But on the other side, it is some kind of contract with (lets say) user of the library. He don't want to shut himself in leg, so he know, he shouldn't use stuff from `testing` module. If he wants to shut himself, why should we protect him from this??
We can even hide this module in docs.

Matcher is public in Market not for tests. In some endpoints I expected to call it directly. But it is possible, that there are some places, where something is public because of tests :(

Reply:
You can use `pub (crate)` to use `Matcher` in endpoints. Now it is only for integration tests.

Meanwhile I was able to remove `testing` module all together.